Proofreading and suggestions (#5)
eosnetworkfoundation/grant-framework: applications/oci-transaction-lifecycle.md

## Project Overview

This project is in response to the API Blue Paper section titled: API Transaction Lifecycle.

### Overview

- **Name:** API Transaction Lifecycle
- **Brief Description:** The API Transaction Lifecycle project is designed to provide EOSIO users clear visibility into transaction status at each step in the process. Four key enhancements to the EOSIO API code are included in this project: a) Transaction Retry, b) Transaction Finality API, c) Transaction Resource Cost Estimation, d) Subjective Billing Improvements.
- **Relationship to EOSIO:** The core EOSIO APIs to be improved through this project are critical components of the EOSIO code that allows applications to interact with the blockchain.
- **Reason for Interest:** The OCI team has been involved in core EOSIO development since 2017 and has expertise in this area. These enhancements will increase the usability and functionality of the APIs.

### Project Details

> - **A. Transaction Retry:** API nodes will monitor transactions as they enter the system and ensure they are resubmitted into the system if they are not processed in a configurable time period. This feature will establish a pool of known incoming transactions and monitor their inclusion into the blockchain. Once the system reaches an acceptable level of confidence that a transaction has been included in a block, the transaction can be pruned from the pool. If a transaction is identified as missing from the blockchain based on the given criteria, the Transaction Retry feature will attempt to resubmit the transaction to the network for inclusion in future blocks until the point at which the transaction expires.

Author: I wasn't sure what noun "it" referred to in "until the point it expires." I assumed it was the transaction. But it might also have meant the time period. So please make sure I edited this correctly.

Reviewer: you are correct,"it" was transaction

- **B. Transaction Finality Status:** API nodes will monitor transactions as they enter the system and provide a new API method to report transaction status.

> - **C. Transaction Resource Cost Estimation:** Nodes will provide an estimate of resources (central processing unit (CPU), random access memory (RAM), and internet bandwidth (NET)) to perform a transaction when a transaction is sent to a new compute_transaction endpoint. The transaction will be applied subjectively to the local chain to calculate resource costs like CPU and NET and determine the resulting deltas in RAM usage. The response will be returned to the client to inform them of the costs associated with the given transaction data.

Author: Should this one also start with "API nodes" like the others, or are these a different type of node?

Reviewer: I'll correct add API at the start in my next edit.

> - **D. Subjective Billing Improvements:** Nodes will provide an added leeway per account that allows a more permissive “subjective billing” rate (rate of utilization) in order to provide a better user experience while maintaining system integrity. The existing subjective CPU decay window will be made configurable from the existing hard-coded value of 24 hours to allow node operators to provide a more relaxed subjective CPU penalty to users. The existing `disable-subjective-account-billing` function will be expanded to also apply to the 3-strike rule, allowing block producers to prevent abnormal transaction loss for accounts expected to sometimes fail during production.

Author: Same question as C. Should it start with "API nodes" for consistency?

Reviewer: yes and will fix in next change.

> 

Author: Add John Schultz title in parentheses

Change Contact to Dan Fedj

### Team Experience

> Multiple members of the OCI team have worked on the EOSIO code base. Kevin Heifner is one of the top contributors in the EOSIO repo and has worked as a contractor for block.one since 2017. The OCI team has been involved in almost all development aspects of the EOSIO code base.

Author: ```suggestion
The OCI team has been involved in almost all development aspects of the EOSIO code base since 2017. Multiple members of the OCI team have worked on the EOSIO code base, and Brian Johnson and Kevin Heifner are top contributors to the EOSIO repo. 
```

## Development Status

> 

Author: Add LinkedIn profile for Huang-Ming's LinkedIn profile

> If you've already started implementing your project or it is part of a larger repository, please provide a link and a description of the code here. In any case, please provide some documentation on the research and other work you have conducted before applying. This could be:

Author: I assumed this section was meant to be deleted before submission.

> - conversations related to this project with stakeholders from the EOS Network Foundation (Yves La Rose discussed).

Author: This looked like it was a note that needed to be expanded upon before submission

> | Number | Deliverable | Specification |

Author: I noticed that the overview FTE spelled out "three" and then repeated it in parentheses as a numeral, but none of the other sections did this. Was there a reason that this was the only section that spelled out a numeral?

| 2. | EOSIO Subjective Billing Improvements | Nodes will provide an added leeway per account that allows a more permissive “subjective billing” rate (rate of utilization) in order to provide a better user experience while maintaining system integrity. The existing subjective CPU decay window will be made configurable from the existing hard-coded value of 24 hours to allow node operators to provide a more relaxed subjective CPU penalty to their users. The existing `disable-subjective-account-billing` function will be expanded to also apply to the 3-strike rule, allowing block producers to prevent abnormal transaction loss for accounts expected to sometimes fail during production. |  

### Milestone 3 - Project Completion

- **Estimated Duration:** 1 week
- **FTE:**  2
- **Costs:** $80,000 USD

> 

Author: Should there be a brief explanation of what completion activities occur during this week?

- **Estimated Duration:** 1 week
- **FTE:**  2
- **Costs:** $80,000 USD

## Future Plans

> Please include here

Author: This looked like it was meant to be removed before submission.
